src/yul_parser.rs

```diff
@@ -360,19 +422,59 @@ impl Block {
     }
 }
 
-pub fn parse_block(source: &str) -> Result<Block, String> {
-    match BlockParser::parse(Rule::block, source) {
+impl Root {
+    fn from(pair: Pair<Rule>, next_identifier: &mut u64) -> Root {
+        let span = pair.clone().into_span();
+        let location = Some(SourceLocation {
+            start: span.start(),
+            end: span.end(),
+        });
+
+        let mut token_iter = pair.into_inner();
+        let p = token_iter.next().unwrap();
+        let inner = match p.as_rule() {
+            Rule::object => InnerRoot::Object(Object::from(p, next_identifier)),
+            Rule::block => InnerRoot::Block(Block::from(p, next_identifier)),
+            _ => unreachable!(),
+        };
+
+        Root { inner, location }
+    }
+}
+
+pub fn parse_root(source: &str) -> Result<Root, String> {
+    match YulParser::parse(Rule::root, source) {
         Ok(mut pairs) => {
             let mut next_identifier = 1u64;
-            Ok(Block::from(pairs.next().unwrap(), &mut next_identifier))
+            Ok(Root::from(pairs.next().unwrap(), &mut next_identifier))
         }
         Err(error) => Err(format!("{}", error)),
     }
 }
 
+pub fn parse_block(source: &str) -> Result<Block, String> {
```

Comment:
If we use `yulParser::parse(Rule::block`, we might get better error messages here.

Reply:
Ah, but then we have to add soi eoi and stuff. hm.

Reply:
I can add that back if you want